File: evolution.py
import string
import random

TARGET = 'Me thinks it is like a weasel'
RESULT_STRING = ''
ALPHABET = ' ' + string.ascii_letters
CHANCE_OF_MUTATION = 10
GENERATIONS = 100

def generateRandomString():

    global TARGET
    global RESULT_STRING
    global CHANCE_OF_MUTATION

    for i in range(len(TARGET)):
        # Sets a random char from the alphabet to each position of the string
        RESULT_STRING = RESULT_STRING + (ALPHABET[random.randint(0, len(TARGET))])

def mutate():

    global RESULT_STRING
    global ALPHABET
    global CHANCE_OF_MUTATION
    global SCORE

    # Mutates according to its mutation ratio
    if random.random() < CHANCE_OF_MUTATION:
        listOfString = list(RESULT_STRING) # Slices the RESULT_STRING
        listOfAlphabet = list(ALPHABET) # Slices the ALPHABET

        # Selects a random int between the length of the result string
        randomIntFromString = random.randint(0, len(listOfString) - 1)

        # Chooses a char from the string according to the int it chose
        randomCharFromString = listOfString[randomIntFromString]

        # Verifies if the characters in the same position are equal, mutates it to a random character from the alphabet
        # then sets the RESULT_STRING
        if randomCharFromString != TARGET[randomIntFromString]:
            listOfString[randomIntFromString] = listOfAlphabet[random.randint(0, len(listOfAlphabet) - 1)]
            RESULT_STRING = "".join(listOfString)

def generations():

    generateRandomString() # Start a random string for it to mutate
    generation = 1 # Counts the generations

    # While the strings are diferent, mutate!
    while RESULT_STRING != TARGET:
        mutate()
        print(RESULT_STRING + f" - Generation {generation}") # Prints the mutation + generation
        generation += 1 # Increases the generation by one

if __name__ == "__main__":
    generations()

File: evolution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

char ALPHABET[] =
        {
                ' ', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
                'k', 'l', 'm', 'n', 'o','p', 'q', 'r', 's', 't',
                'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D',
                'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
                'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X',
                'Y', 'Z'
        };
int CHANCE_OF_MUTATION = 10;

void generatesRandomString(char* target, char* resultString){
    char ch;
    for (int i = 0; i < strlen(target); i++){
        ch = ALPHABET[rand() % sizeof ALPHABET];
        resultString[i] = ch;
    }
}

void mutate(char* target, char* resultString){
    int randomInt = rand() % 100;
    int randomIntFromString = rand() % (strlen(resultString) + 1);
    char randomCharFromString = resultString[randomIntFromString];

    if (randomInt < CHANCE_OF_MUTATION){
        if (resultString[randomIntFromString] != target[randomIntFromString]){
            resultString[randomIntFromString] = ALPHABET[rand() % strlen(ALPHABET)];
        }
    }
}
int main() {
    time_t t;
    char TARGET[] = "Me thinks it is like a weasel ";
    char RESULT_STRING[28];
    int generation = 1;

    srand((unsigned ) time(&t));
    generatesRandomString(TARGET, RESULT_STRING);

    clock_t start = clock();
    while(strcmp(RESULT_STRING, TARGET) != 0){
        mutate(TARGET, RESULT_STRING);
        printf("%s - Generation %d\n", RESULT_STRING, generation);
        generation++;
    }
    clock_t end = clock();
    printf("Time: %f - miliseconds", (double)(end - start) * 1000/CLOCKS_PER_SEC );

    return 0;
}
